### PR TITLE
Components provider

### PR DESCRIPTION
**Pull request by webrouse**
Changes:
- created `components.Provider`
- removed hard dependency of the `local` package on the `remote.StorageApi` ... instead of the API, `local` package now depends on the `components.Provider`.

### CODE REVIEW

File: src/components/provider.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	"keboola-as-code/src/model"
+	"sort"
+	"sync"
+)
+
+type remoteProvider interface {
+	GetComponent(componentId string) (*model.Component, error)
+}
+
+type Provider struct {
```
Comment (webrouse) [inline src/components/provider.go:14]: Namiesto `remote.ComponentsCache` je `components.Provider` ... tym ze som pouzil interface `remoteProvider`, tak uz tu nie je zavislost na `remote.StorageAPI`.

Problem bol, ze informacie o komponentach potrebujem aj v `local` package, ... a v `remote` zase potrebujem `local` na ulozenie noveho ID na disk. Takze vznikala zavislost `local` -> `remote` -> `local` ... a teraz to je `local` -> `components`, ... co je viac logicke a nerobi to problemy.

File: src/remote/component.go [hunk 2]
```diff
@@ -29,62 +26,3 @@ func (a *StorageApi) GetComponentRequest(componentId string) *client.Request {
 			a.Components().Set(component)
 		})
 }
-
-type ComponentsCache struct {
```
Comment (webrouse) [inline src/remote/component.go:33]: Presunute.

**Review by webrouse — COMMENTED** [review]
Vznikala mi cyklicka zavilost `local` -> `remote` -> `local`, vyriesil som to vytvorenim `components.Provider`, a teda teraz, `local` -> `components` a `remote` -> `local`.

File: src/utils/error.go [hunk 3]
```diff
@@ -22,23 +24,29 @@ func (e *ErrorRaw) Error() string {
...
 func NewMultiError() *Error {
-	e := &Error{multiError: &multierror.Error{}}
+	e := &Error{multiError: &multierror.Error{}, lock: &sync.Mutex{}}
 	e.ErrorFormat = formatError
 	return e
 }
 
 // Append error
 func (e *Error) Append(err error) {
+	e.lock.Lock()
```
Comment (webrouse) [inline src/utils/error.go:34]: Okrem toho som narazil na chybajuci zamok, ... ak nastane naraz viacero chyb z viacerych `goroutines`.

File: src/state/state.go [hunk 4]
```diff
@@ -248,6 +249,9 @@ func (s *State) SetLocalState(local model.Object, record model.Record) model.Obj
 }
 
 func (s *State) getOrCreate(key model.Key) (model.ObjectState, error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
```
Comment (webrouse) [inline src/state/state.go:253]: Tu chybal zamok -> `data race`.

File: src/local/manager.go [hunk 5]
```diff
@@ -2,22 +2,22 @@ package local
...
-	"keboola-as-code/src/remote"
 )
 
 type Manager struct {
-	logger   *zap.SugaredLogger
-	manifest *manifest.Manifest
-	api      *remote.StorageApi
+	logger     *zap.SugaredLogger
+	manifest   *manifest.Manifest
+	components *components.Provider
```
Comment (webrouse) [inline src/local/manager.go:13]: `local` package uz nezavisi na `remote` ale na `components`.

**Review by kacurez — APPROVED** [review]
lgtm

**Pull request merged**